src/CTA.Rules.Metrics/PortSolutionResultReportGenerator.cs

```diff
@@ -45,20 +48,32 @@ private void GenerateMetrics()
             DownloadedFilesMetrics = MetricsTransformer.TransformDownloadedFiles(Context, PortSolutionResult.DownloadedFiles);
 
             // Gather project-level metrics
+            var targetVersionMetrics = new List<TargetVersionMetric>();
+            var upgradePackageMetrics = new List<UpgradePackageMetric>();
+            var actionPackageMetrics = new List<ActionPackageMetric>();
             var actionExecutionMetrics = new List<GenericActionExecutionMetric>();
+
             foreach (var projectResult in PortSolutionResult.ProjectResults)
             {
-                var executedActionMetrics = MetricsTransformer.TransformGenericActionExecutions(Context, projectResult);
-                actionExecutionMetrics.AddRange(executedActionMetrics);
+                targetVersionMetrics.AddRange(MetricsTransformer.TransformTargetVersions(Context, projectResult));
+                upgradePackageMetrics.AddRange(MetricsTransformer.TransformUpgradePackages(Context, projectResult));
+                actionPackageMetrics.AddRange(MetricsTransformer.TransformActionPackages(Context, projectResult));
+                actionExecutionMetrics.AddRange(MetricsTransformer.TransformGenericActionExecutions(Context, projectResult));
             }
+            TargetVersionMetrics = targetVersionMetrics;
+            UpgradePackageMetrics = upgradePackageMetrics;
+            ActionPackageMetrics = actionPackageMetrics;
             GenericActionExecutionMetrics = actionExecutionMetrics;
-        }
+        } 
 
         private string GeneratePortSolutionResultJsonReport()
         {
             var allMetrics = new List<CTAMetric>();
             allMetrics = allMetrics.Concat(ReferencesMetrics)
                 .Concat(DownloadedFilesMetrics)
+                .Concat(TargetVersionMetrics)
```

Comment:
ensure with additional metrics going as json, there is no performance bottleneck... meaning when we are uploading these json, all those should be async way... else see if you can point me the code where it's being uploaded to s3..

Reply:
These metrics are stored in log messages which are processed by the Porting Assistant logger, so the way these metrics are written to disk and stored to S3 are the same as how Porting Assistant uploads logs to S3... Unfortunately, I'm not sure where in the PA code that is done 🙂